ably/realtime_channel.go

```diff
@@ -203,34 +252,68 @@ func (c *RealtimeChannel) mayAttach(checkActive bool) (result, error) {
 	if checkActive {
 		switch c.state {
 		case ChannelStateAttaching:
-			return c.internalEmitter.listenResult(ChannelStateAttached, ChannelStateFailed), nil
-		case ChannelStateAttached:
+			return c.internalEmitter.listenResult(ChannelStateAttached, ChannelStateDetached, ChannelStateSuspended, ChannelStateFailed), nil // RTL4h
+		case ChannelStateAttached: // RTL4a
 			return nopResult, nil
 		}
 	}
 	return c.lockAttach(nil)
 }
 
 func (c *RealtimeChannel) lockAttach(err error) (result, error) {
+	if c.state == ChannelStateFailed { // RTL4g
+		err = nil
+	}
+
 	switch c.client.Connection.State() {
 	// RTL4b
 	case ConnectionStateInitialized,
 		ConnectionStateClosed,
 		ConnectionStateClosing,
+		ConnectionStateSuspended,
 		ConnectionStateFailed:
-		return nil, newError(80000, errAttach)
+		return nil, newError(ErrConnectionFailed, errConnAttach(c.client.Connection.State()))
```

Comment:
Haha, not sure, seems I have followed it from the originalCode, I vaguely remember the reason was, I wasn't able to 
find the error code for `errChannelAttachFailed`, we don't have error code for channel attach : (

Reply:
I believe, it's because we are doing explicit check on connection states and since connections are in invalid states, we are specifying error as `ErrConnectionFailed`